beets/dbcore/query.py

```diff
@@ -550,19 +551,38 @@ def parse(cls, string):
         """Parse a date and return a `Period` object or `None` if the
         string is empty.
         """
-        if not string:
-            return None
-        ordinal = string.count('-')
-        if ordinal >= len(cls.date_formats):
-            # Too many components.
-            return None
-        date_format = cls.date_formats[ordinal]
-        try:
-            date = datetime.strptime(string, date_format)
-        except ValueError:
-            # Parsing failed.
-            return None
-        precision = cls.precisions[ordinal]
+        pattern_dq = '@([+|-]?)(\d+)([y|m|w|d])'
+        if re.match(pattern_dq, string) is not None:
+            sign = re.match(pattern_dq, string).group(1)
+            quantity = re.match(pattern_dq, string).group(2)
+            timespan = re.match(pattern_dq, string).group(3)
+            if sign == '-':
+                m = -1
+            else:
+                m = 1
+            if timespan == 'y':
+                date = datetime.now() + m * timedelta(days=int(quantity) * 365)
+            elif timespan == 'm':
+                date = datetime.now() + m * timedelta(days=int(quantity) * 30)
+            elif timespan == 'w':
+                date = datetime.now() + m * timedelta(days=int(quantity) * 7)
+            elif timespan == 'd':
+                date = datetime.now() + m * timedelta(days=int(quantity))
```

Comment:
done, I used the relative tuple that already existed, just "upgraded" it to a dict